package com.hartwig.serve.transvar;

import static com.hartwig.serve.transvar.util.Nucleotides.reverseStrandBases;
import static com.hartwig.serve.util.AminoAcids.findAminoAcidForCodon;

import java.io.File;
import java.io.FileNotFoundException;
import java.util.List;

import com.google.common.annotations.VisibleForTesting;
import com.google.common.collect.Lists;
import com.hartwig.serve.common.RefGenomeFunctions;
import com.hartwig.serve.common.ensemblcache.Strand;
import com.hartwig.serve.datamodel.RefGenome;
import com.hartwig.serve.extraction.hotspot.Hotspot;
import com.hartwig.serve.extraction.hotspot.ImmutableHotspot;
import com.hartwig.serve.transvar.datamodel.TransvarAnnotation;
import com.hartwig.serve.transvar.datamodel.TransvarComplexInsertDelete;
import com.hartwig.serve.transvar.datamodel.TransvarDeletion;
import com.hartwig.serve.transvar.datamodel.TransvarDuplication;
import com.hartwig.serve.transvar.datamodel.TransvarFrameshift;
import com.hartwig.serve.transvar.datamodel.TransvarInsertion;
import com.hartwig.serve.transvar.datamodel.TransvarRecord;
import com.hartwig.serve.transvar.datamodel.TransvarSnvMnv;
import com.hartwig.serve.transvar.util.Nucleotides;
import com.hartwig.serve.util.AminoAcids;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import htsjdk.samtools.reference.IndexedFastaSequenceFile;

class TransvarInterpreter {

    private static final Logger LOGGER = LogManager.getLogger(TransvarInterpreter.class);

    @NotNull
    private final RefGenome refGenome;
    @NotNull
    private final IndexedFastaSequenceFile refGenomeFasta;

    @NotNull
    static TransvarInterpreter withRefGenome(@NotNull RefGenome refGenome, @NotNull String refGenomeFastaFile)
            throws FileNotFoundException {
        return new TransvarInterpreter(refGenome, new IndexedFastaSequenceFile(new File(refGenomeFastaFile)));
    }

    private TransvarInterpreter(@NotNull final RefGenome refGenome, @NotNull final IndexedFastaSequenceFile refGenomeFasta) {
        this.refGenome = refGenome;
        this.refGenomeFasta = refGenomeFasta;
    }

    @NotNull
    List<Hotspot> convertRecordToHotspots(@NotNull TransvarRecord record, @NotNull Strand strand) {
        TransvarAnnotation annotation = record.annotation();
        if (annotation instanceof TransvarSnvMnv) {
            return convertSnvMnvToHotspots(record, (TransvarSnvMnv) annotation, strand);
        } else if (annotation instanceof TransvarDuplication) {
            return convertDuplicationToHotspots(record, (TransvarDuplication) annotation);
        } else if (annotation instanceof TransvarInsertion) {
            return convertInsertionToHotspots(record, (TransvarInsertion) annotation, strand);
        } else if (annotation instanceof TransvarDeletion) {
            return convertDeletionToHotspots(record, (TransvarDeletion) annotation);
        } else if (annotation instanceof TransvarComplexInsertDelete) {
            return convertComplexInsertDeleteToHotspots(record, (TransvarComplexInsertDelete) annotation, strand);
        } else if (annotation instanceof TransvarFrameshift) {
            return convertFrameshiftToHotspots(record, (TransvarFrameshift) annotation, strand);
        } else {
            LOGGER.warn("Unrecognized annotation type in transvar record: '{}'. Skipping interpretation.",
                    annotation.getClass().toString());
            return Lists.newArrayList();
        }
    }

    @NotNull
    private List<Hotspot> convertSnvMnvToHotspots(@NotNull TransvarRecord record, @NotNull TransvarSnvMnv snvMnv, @NotNull Strand strand) {
        List<Hotspot> hotspots = Lists.newArrayList();

        if (record.variantSpanMultipleExons()) {
            // In this case we only generate hotspots for the base mutation in case it is an SNV.
            if (snvMnv.gdnaRef().length() == 1) {
                hotspots.add(withRefBasedChromosome(record.chromosome()).position(record.gdnaPosition())
                        .ref(snvMnv.gdnaRef())
                        .alt(snvMnv.gdnaAlt())
                        .build());
            }
        } else {
            // We need to look up which index of the ref codon is changed (0, 1 or 2) in case of SNV/MNV.
            Integer gdnaCodonIndex = findIndexInRefCodonForGdnaMatch(snvMnv, strand);

            if (gdnaCodonIndex != null) {
                for (String candidateCodon : snvMnv.candidateCodons()) {
                    hotspots.add(fromCandidateCodon(record, snvMnv.referenceCodon(), candidateCodon, gdnaCodonIndex, strand));
                }
            } else {
                LOGGER.warn("Could not resolve gdnaCodonIndex for '{}' on strand {}", record, strand);
            }
        }

        return hotspots;
    }

    @Nullable
    private static Integer findIndexInRefCodonForGdnaMatch(@NotNull TransvarSnvMnv snvMnv, @NotNull Strand strand) {
        String codonCompatibleRef = strand.equals(Strand.FORWARD) ? snvMnv.gdnaRef() : reverseStrandBases(snvMnv.gdnaRef());
        String codonCompatibleAlt = strand.equals(Strand.FORWARD) ? snvMnv.gdnaAlt() : reverseStrandBases(snvMnv.gdnaAlt());

        // Function only supports SNV and MNV
        assert codonCompatibleRef.length() == codonCompatibleAlt.length();

        // We look for the candidate codon where the mutation is exclusively the mutation implied by the ref>alt
        int mutLength = codonCompatibleRef.length();
        for (String candidateCodon : snvMnv.candidateCodons()) {
            for (int i = 0; i < 4 - mutLength; i++) {
                if (snvMnv.referenceCodon().substring(i, i + mutLength).equals(codonCompatibleRef) && candidateCodon.substring(i,
                        i + mutLength).equals(codonCompatibleAlt)) {
                    boolean match = true;
                    for (int j = 0; j < 3; j++) {
                        // No match if we find another mismatch outside the ref->alt range.
                        if ((j - i >= mutLength || j - i < 0) && !snvMnv.referenceCodon()
                                .substring(j, j + 1)
                                .equals(candidateCodon.substring(j, j + 1))) {
                            match = false;
                        }
                    }
                    if (match) {
                        // For reverse strand searches we need to correct for the flipping of MNVs.
                        return strand == Strand.FORWARD ? i : i + mutLength - 1;
                    }
                }
            }
        }

        return null;
    }

    @NotNull
    private Hotspot fromCandidateCodon(@NotNull TransvarRecord record, @NotNull String referenceCodon, @NotNull String candidateCodon,
            int gdnaCodonIndex, @NotNull Strand strand) {
        String strandAdjustedRefCodon = strand == Strand.FORWARD ? referenceCodon : reverseStrandBases(referenceCodon);
        String strandAdjustedCandidateCodon = strand == Strand.FORWARD ? candidateCodon : reverseStrandBases(candidateCodon);
        int strandAdjustedGdnaCodingIndex = strand == Strand.FORWARD ? gdnaCodonIndex : 2 - gdnaCodonIndex;

        int firstMutatedPosition = -1;
        int lastMutatedPosition = -1;
        for (int i = 0; i < 3; i++) {
            if (!strandAdjustedRefCodon.substring(i, i + 1).equals(strandAdjustedCandidateCodon.substring(i, i + 1))) {
                if (firstMutatedPosition == -1) {
                    firstMutatedPosition = i;
                }
                lastMutatedPosition = i;
            }
        }

        String ref = strandAdjustedRefCodon.substring(firstMutatedPosition, lastMutatedPosition + 1);
        String alt = strandAdjustedCandidateCodon.substring(firstMutatedPosition, lastMutatedPosition + 1);

        return withRefBasedChromosome(record.chromosome()).position(
                record.gdnaPosition() - strandAdjustedGdnaCodingIndex + firstMutatedPosition).ref(ref).alt(alt).build();
    }

    @NotNull
    private List<Hotspot> convertDuplicationToHotspots(@NotNull TransvarRecord record, @NotNull TransvarDuplication dup) {
        List<Hotspot> hotspots = Lists.newArrayList();
        if (!record.variantSpanMultipleExons()) {
            // Dups don't have ref and alt information so need to look it up in ref genome.
            int position = record.gdnaPosition() - 1;
            String preMutatedSequence = refSequence(record.chromosome(), position, position);
            String dupBases = refSequence(record.chromosome(), position + 1, position + dup.duplicatedBaseCount());

            hotspots.add(withRefBasedChromosome(record.chromosome()).position(position)
                    .ref(preMutatedSequence)
                    .alt(preMutatedSequence + dupBases)
                    .build());
        } else {
            LOGGER.debug("Duplication spanning multiple exons. Ignoring '{}'", record);
        }

        return hotspots;
    }

    @NotNull
    private List<Hotspot> convertInsertionToHotspots(@NotNull TransvarRecord record, @NotNull TransvarInsertion insertion,
            @NotNull Strand strand) {
        List<Hotspot> hotspots = Lists.newArrayList();

        int position = record.gdnaPosition();
        String preMutatedSequence = refSequence(record.chromosome(), position, position);

        ImmutableHotspot.Builder hotspotBuilder = withRefBasedChromosome(record.chromosome()).position(position).ref(preMutatedSequence);

        // We assume inserts of length 3 are always (inframe) amino acid inserts,
        //  and we know the inserted bases match the proteins inserted.
        if (insertion.insertedBases().length() == 3) {
            for (String trinucleotide : allTrinucleotidesForSameAminoAcid(insertion.insertedBases(), strand)) {
                hotspots.add(hotspotBuilder.alt(preMutatedSequence + trinucleotide).build());
            }
        } else {
            hotspots.add(hotspotBuilder.alt(preMutatedSequence + insertion.insertedBases()).build());
        }

        return hotspots;
    }

    @NotNull
    public static List<String> allTrinucleotidesForSameAminoAcid(@NotNull String trinucleotideToFind, @NotNull Strand strand) {
        if (trinucleotideToFind.length() != 3) {
            LOGGER.warn("Cannot look up amino acids for non-trinucleotides: {}", trinucleotideToFind);
            return Lists.newArrayList();
        }

        String trinucleotideBases = strand == Strand.FORWARD ? trinucleotideToFind : reverseStrandBases(trinucleotideToFind);
        String aminoAcid = findAminoAcidForCodon(trinucleotideBases);

        List<String> allTrinucleotides = Lists.newArrayList();
        List<String> trinucleotides = AminoAcids.AMINO_ACID_TO_CODON_MAP.get(aminoAcid);
        if (trinucleotides != null) {
            for (String trinucleotide : trinucleotides) {
                allTrinucleotides.add(strand == Strand.FORWARD ? trinucleotide : reverseStrandBases(trinucleotide));
            }
        } else {
            LOGGER.warn("Could not find amino acid for trinucleotide '{}' on {} strand", trinucleotideToFind, strand);
        }

        return allTrinucleotides;
    }

    @NotNull
    private List<Hotspot> convertDeletionToHotspots(@NotNull TransvarRecord record, @NotNull TransvarDeletion deletion) {
        List<Hotspot> hotspots = Lists.newArrayList();
        if (!record.variantSpanMultipleExons()) {
            ImmutableHotspot.Builder hotspotBuilder = withRefBasedChromosome(record.chromosome());
            for (int start = deletion.leftAlignedGDNAPosition(); start <= record.gdnaPosition(); start++) {
                int adjustedPosition = start - 1;
                String preMutatedSequence = refSequence(record.chromosome(), adjustedPosition, adjustedPosition);
                String deletedSequence =
                        refSequence(record.chromosome(), adjustedPosition + 1, adjustedPosition + deletion.deletedBaseCount());

                hotspots.add(hotspotBuilder.position(adjustedPosition)
                        .ref(preMutatedSequence + deletedSequence)
                        .alt(preMutatedSequence)
                        .build());
            }
        } else {
            LOGGER.debug("Deletion spanning multiple exons. Ignoring '{}'", record);
        }

        return hotspots;
    }

    @NotNull
    private List<Hotspot> convertComplexInsertDeleteToHotspots(@NotNull TransvarRecord record, @NotNull TransvarComplexInsertDelete insDel,
            @NotNull Strand strand) {
        List<Hotspot> hotspots = Lists.newArrayList();
        if (!record.variantSpanMultipleExons()) {
            int position = record.gdnaPosition();
            String deletedBases = refSequence(record.chromosome(), position, position + insDel.deletedBaseCount() - 1);

            ImmutableHotspot.Builder hotspotBuilder = withRefBasedChromosome(record.chromosome()).position(position).ref(deletedBases);

            String insertedBases = insDel.insertedSequence();
            hotspots.add(reduceComplexityForComplexInsDel(hotspotBuilder.alt(insertedBases).build()));
            // For now, we only add alternative sequences for insertions of one AA
            if (insertedBases.length() == 3) {
                for (String candidateAlternativeCodon : insDel.candidateAlternativeCodons()) {
                    assert candidateAlternativeCodon.length() == insertedBases.length();
                    String strandAdjustedAlternativeSequence =
                            strand == Strand.FORWARD ? candidateAlternativeCodon : reverseStrandBases(candidateAlternativeCodon);
                    if (!strandAdjustedAlternativeSequence.equals(insertedBases)) {
                        hotspots.add(reduceComplexityForComplexInsDel(hotspotBuilder.alt(strandAdjustedAlternativeSequence).build()));
                    }
                }
            }
        } else {
            LOGGER.debug("Complex insert/delete spanning multiple exons. Ignoring '{}'", record);
        }
        return hotspots;
    }

    @NotNull
    @VisibleForTesting
    Hotspot reduceComplexityForComplexInsDel(@NotNull Hotspot complexInsDel) {
        assert complexInsDel.ref().length() > 1 && complexInsDel.alt().length() > 1;

        String simplifiedRef = complexInsDel.ref();
        String simplifiedAlt = complexInsDel.alt();
        while (simplifiedRef.length() > 1 && simplifiedAlt.length() > 1
                && simplifiedRef.charAt(simplifiedRef.length() - 1) == simplifiedAlt.charAt(simplifiedAlt.length() - 1)) {
            simplifiedRef = simplifiedRef.substring(0, simplifiedRef.length() - 1);
            simplifiedAlt = simplifiedAlt.substring(0, simplifiedAlt.length() - 1);
        }

        int adjustedPos = complexInsDel.position();
        while (simplifiedRef.length() > 2 && simplifiedAlt.length() > 2 && simplifiedRef.subSequence(0, 1)
                .equals(simplifiedAlt.substring(0, 1))) {
            adjustedPos++;
            simplifiedRef = simplifiedRef.substring(1);
            simplifiedAlt = simplifiedAlt.substring(1);
        }

        return withRefBasedChromosome(complexInsDel.chromosome()).position(adjustedPos).ref(simplifiedRef).alt(simplifiedAlt).build();
    }

    @NotNull
    private List<Hotspot> convertFrameshiftToHotspots(@NotNull TransvarRecord record, @NotNull TransvarFrameshift frameshift,
            @NotNull Strand strand) {
        List<Hotspot> hotspots = Lists.newArrayList();
        if (!record.variantSpanMultipleExons()) {
            int posPriorToCodon = strand == Strand.FORWARD ? record.gdnaPosition() : record.gdnaPosition() - 3;
            // For frameshifts in start codons, transvar generates the start of the start codon rather than position prior.
            if (frameshift.isFrameshiftInsideStartCodon()) {
                posPriorToCodon = strand == Strand.FORWARD ? posPriorToCodon - 1 : posPriorToCodon + 1;
            }
            String referenceCodon = refSequence(record.chromosome(), posPriorToCodon + 1, posPriorToCodon + 3);
            String refAminoAcid = findAminoAcidForCodon(strand == Strand.FORWARD ? referenceCodon : reverseStrandBases(referenceCodon));

            if (refAminoAcid == null) {
                LOGGER.warn("Could not resolve a valid ref amino acid for '{}' based on reference codon {}. Skipping hotspot generation.",
                        record,
                        referenceCodon);
                return Lists.newArrayList();
            }

            hotspots.addAll(generateSingleBaseInserts(record, posPriorToCodon, strand, refAminoAcid));
            hotspots.addAll(generateSingleBaseDeletes(record, posPriorToCodon, strand, refAminoAcid));
            hotspots.addAll(generateDoubleBaseDeletes(record, posPriorToCodon, strand, refAminoAcid));
        } else {
            LOGGER.debug("Frameshift spanning multiple exons. Ignoring '{}'", record);
        }

        return hotspots;
    }

    @NotNull
    private List<Hotspot> generateSingleBaseInserts(@NotNull TransvarRecord record, int posPriorToCodon, @NotNull Strand strand,
            @NotNull String refAminoAcid) {
        List<Hotspot> hotspots = Lists.newArrayList();
        ImmutableHotspot.Builder builder = withRefBasedChromosome(record.chromosome());

        // Add 12 single base insertions in case they don't lead to synonymous impact in the impacted codon
        for (int i = 0; i < 3; i++) {
            // For reverse strand we need to move the position up by 1 for inserts
            int strandCorrection = strand == Strand.FORWARD ? 0 : 1;
            int pos = posPriorToCodon + i + strandCorrection;
            String ref = refSequence(record.chromosome(), pos, pos);
            builder.position(pos).ref(ref);

            String refBase1 =
                    refSequence(record.chromosome(), posPriorToCodon + 1 + strandCorrection, posPriorToCodon + 1 + strandCorrection);
            String refBase2 =
                    refSequence(record.chromosome(), posPriorToCodon + 2 + strandCorrection, posPriorToCodon + 2 + strandCorrection);
            for (char base : Nucleotides.DNA_BASES) {
                String newRefCodon;
                if (i == 0) {
                    newRefCodon = base + refBase1 + refBase2;
                } else if (i == 1) {
                    newRefCodon = refBase1 + base + refBase2;
                } else {
                    newRefCodon = refBase1 + refBase2 + base;
                }
                String newAminoAcid = findAminoAcidForCodon(strand == Strand.FORWARD ? newRefCodon : reverseStrandBases(newRefCodon));

                if (newAminoAcid != null && !newAminoAcid.equals(refAminoAcid)) {
                    hotspots.add(builder.alt(ref + base).build());
                }
            }
        }
        return hotspots;
    }

    @NotNull
    private List<Hotspot> generateSingleBaseDeletes(@NotNull TransvarRecord record, int posPriorToCodon, @NotNull Strand strand,
            @NotNull String refAminoAcid) {
        List<Hotspot> hotspots = Lists.newArrayList();
        ImmutableHotspot.Builder builder = withRefBasedChromosome(record.chromosome());

        // Add the 3 single base deletes in case they don't lead to synonymous impact in the impacted codon
        for (int i = 0; i < 3; i++) {
            String newRefCodon;
            if (strand == Strand.FORWARD) {
                if (i == 0) {
                    newRefCodon = refSequence(record.chromosome(), posPriorToCodon + 2, posPriorToCodon + 4);
                } else if (i == 1) {
                    newRefCodon =
                            refSequence(record.chromosome(), posPriorToCodon + 1, posPriorToCodon + 1) + refSequence(record.chromosome(),
                                    posPriorToCodon + 3,
                                    posPriorToCodon + 4);
                } else {
                    newRefCodon =
                            refSequence(record.chromosome(), posPriorToCodon + 1, posPriorToCodon + 2) + refSequence(record.chromosome(),
                                    posPriorToCodon + 4,
                                    posPriorToCodon + 4);
                }
            } else {
                if (i == 0) {
                    newRefCodon = refSequence(record.chromosome(), posPriorToCodon, posPriorToCodon) + refSequence(record.chromosome(),
                            posPriorToCodon + 2,
                            posPriorToCodon + 3);
                } else if (i == 1) {
                    newRefCodon = refSequence(record.chromosome(), posPriorToCodon, posPriorToCodon + 1) + refSequence(record.chromosome(),
                            posPriorToCodon + 3,
                            posPriorToCodon + 3);
                } else {
                    newRefCodon = refSequence(record.chromosome(), posPriorToCodon, posPriorToCodon + 2);
                }
            }
            String newAminoAcid = findAminoAcidForCodon(strand == Strand.FORWARD ? newRefCodon : reverseStrandBases(newRefCodon));

            if (newAminoAcid != null && !newAminoAcid.equals(refAminoAcid)) {
                int pos = posPriorToCodon + i;
                String ref = refSequence(record.chromosome(), pos, pos + 1);
                String alt = refSequence(record.chromosome(), pos, pos);

                hotspots.add(builder.position(pos).ref(ref).alt(alt).build());
            }
        }
        return hotspots;
    }

    @NotNull
    private List<Hotspot> generateDoubleBaseDeletes(@NotNull TransvarRecord record, int posPriorToCodon, @NotNull Strand strand,
            @NotNull String refAminoAcid) {
        List<Hotspot> hotspots = Lists.newArrayList();
        ImmutableHotspot.Builder builder = withRefBasedChromosome(record.chromosome());

        // Add the 2 double base deletes in case they don't lead to synonymous impact in the impacted codon
        for (int i = 0; i < 2; i++) {
            String newRefCodon;
            if (strand == Strand.FORWARD) {
                if (i == 0) {
                    newRefCodon = refSequence(record.chromosome(), posPriorToCodon + 3, posPriorToCodon + 5);
                } else {
                    newRefCodon =
                            refSequence(record.chromosome(), posPriorToCodon + 1, posPriorToCodon + 1) + refSequence(record.chromosome(),
                                    posPriorToCodon + 4,
                                    posPriorToCodon + 5);
                }
            } else {
                if (i == 0) {
                    newRefCodon = refSequence(record.chromosome(), posPriorToCodon - 1, posPriorToCodon) + refSequence(record.chromosome(),
                            posPriorToCodon + 3,
                            posPriorToCodon + 3);
                } else {
                    newRefCodon = refSequence(record.chromosome(), posPriorToCodon - 1, posPriorToCodon + 1);
                }
            }

            String newAminoAcid = findAminoAcidForCodon(strand == Strand.FORWARD ? newRefCodon : reverseStrandBases(newRefCodon));

            if (newAminoAcid != null && !newAminoAcid.equals(refAminoAcid)) {
                int pos = posPriorToCodon + i;
                String ref = refSequence(record.chromosome(), pos, pos + 2);
                String alt = refSequence(record.chromosome(), pos, pos);

                hotspots.add(builder.position(pos).ref(ref).alt(alt).build());
            }
        }
        return hotspots;
    }

    @NotNull
    private String refSequence(@NotNull String chromosome, int start, int end) {
        String versionedChromosome = RefGenomeFunctions.versionedChromosome(chromosome, refGenome);
        return refGenomeFasta.getSubsequenceAt(versionedChromosome, start, end).getBaseString();
    }

    @NotNull
    private ImmutableHotspot.Builder withRefBasedChromosome(@NotNull String chromosome) {
        return ImmutableHotspot.builder().chromosome(RefGenomeFunctions.versionedChromosome(chromosome, refGenome));
    }
}
